beacon/light/committee_chain.go

```diff
@@ -0,0 +1,528 @@
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package light
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/beacon/params"
+	"github.com/ethereum/go-ethereum/beacon/types"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/lru"
+	"github.com/ethereum/go-ethereum/common/mclock"
+	"github.com/ethereum/go-ethereum/core/rawdb"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rlp"
+)
+
+var (
+	ErrNeedCommittee      = errors.New("sync committee required")
+	ErrInvalidUpdate      = errors.New("invalid committee update")
+	ErrInvalidPeriod      = errors.New("invalid update period")
+	ErrWrongCommitteeRoot = errors.New("wrong committee root")
+	ErrCannotReorg        = errors.New("can not reorg committee chain")
+)
+
+// CommitteeChain is a passive data structure that can validate, hold and update
+// a chain of beacon light sync committees and updates. It requires at least one
+// externally set fixed committee root at the beginning of the chain which can
+// be set either based on a CheckpointData or a trusted source (a local beacon
+// full node). This makes the structure useful for both light client and light
+// server setups.
+//
+// It always maintains the following consistency constraints:
+//   - a committee can only be present if its root hash matches an existing fixed
+//     root or if it is proven by an update at the previous period
+//   - an update can only be present if a committee is present at the same period
+//     and the update signature is valid and has enough participants.
+//     The committee at the next period (proven by the update) should also be
+//     present (note that this means they can only be added together if neither
+//     is present yet). If a fixed root is present at the next period then the
+//     update can only be present if it proves the same committee root.
+//
+// Once synced to the current sync period, CommitteeChain can also validate
+// signed beacon headers.
+type CommitteeChain struct {
+	// chainmu guards against concurrent access to the canonicalStore structures
+	// (updates, committees, fixedRoots) and ensures that they stay consistent
+	// with each other and with committeeCache.
+	chainmu        sync.RWMutex
+	db             ethdb.KeyValueStore
+	updates        *canonicalStore[*types.LightClientUpdate]
+	committees     *canonicalStore[*types.SerializedSyncCommittee]
+	fixedRoots     *canonicalStore[common.Hash]
```

Comment:
`CommitteeChain` is also used by servers. Actually I want light clients to also be able to serve light sync data (committee chain and latest signed head) to other clients. I'd like to move away from the strict client-server separation and make it a typical use case that some node follows the beacon chain as a light client but then processes and serves the state, or just updates and serves a slice of the state.
Another reason to have the chain even as a client-only node is the hopefully largely theoretical possibility of low sync committee signer participation where we might temporarily accept below 2/3 signers but this means that reorgs are also possible. I hope this will never come up on mainnet but the capability of recovering from such a scenario could make such an attack unfeasible.

Also we can later add a prune feature to `CommitteeChain` so if one really wants to be lightweight then always pruning everything before the latest period is still an option.

Reply:
I agree about "fixedRoots", renamed to "fixedCommitteeRoots".